README.rst
Python Client for Private Certificate Authority API
===================================================

|stable| |pypi| |versions|

`Private Certificate Authority API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-private-ca.svg
   :target: https://pypi.org/project/google-cloud-private-ca/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-private-ca.svg
   :target: https://pypi.org/project/google-cloud-private-ca/
.. _Private Certificate Authority API: https://cloud.google.com/certificate-authority-service
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/privateca/latest
.. _Product Documentation:  https://cloud.google.com/certificate-authority-service

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Private Certificate Authority API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Private Certificate Authority API.:  https://cloud.google.com/certificate-authority-service
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-private-ca


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-private-ca

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Private Certificate Authority API
   to see other available methods on the client.
-  Read the `Private Certificate Authority API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Private Certificate Authority API Product documentation:  https://cloud.google.com/certificate-authority-service
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [1.3.0](https://www.github.com/googleapis/python-security-private-ca/compare/v1.2.1...v1.3.0) (2022-01-08)


### Features

* **samples:** add subordinate CA samples ([94b6801](https://www.github.com/googleapis/python-security-private-ca/commit/94b68019a4d9a83dcada4a01fb33599e3f71a283))

### [1.2.1](https://www.github.com/googleapis/python-security-private-ca/compare/v1.2.0...v1.2.1) (2021-11-02)


### Bug Fixes

* **deps:** drop packaging dependency ([7a1d242](https://www.github.com/googleapis/python-security-private-ca/commit/7a1d2429be86650e03607ca4435fe8a8593509ae))
* **deps:** require google-api-core >= 1.28.0 ([7a1d242](https://www.github.com/googleapis/python-security-private-ca/commit/7a1d2429be86650e03607ca4435fe8a8593509ae))


### Documentation

* fix docstring formatting ([#141](https://www.github.com/googleapis/python-security-private-ca/issues/141)) ([79e9911](https://www.github.com/googleapis/python-security-private-ca/commit/79e991145543f7588d21bcc274eba983aba3653d))
* list oneofs in docstring ([7a1d242](https://www.github.com/googleapis/python-security-private-ca/commit/7a1d2429be86650e03607ca4435fe8a8593509ae))

## [1.2.0](https://www.github.com/googleapis/python-security-private-ca/compare/v1.1.0...v1.2.0) (2021-10-25)


### Features

* add support for python 3.10 ([#135](https://www.github.com/googleapis/python-security-private-ca/issues/135)) ([01d8632](https://www.github.com/googleapis/python-security-private-ca/commit/01d8632c9f974d9e5aafbfcf48475de822600f54))

## [1.1.0](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.6...v1.1.0) (2021-10-08)


### Features

* add context manager support in client ([#131](https://www.github.com/googleapis/python-security-private-ca/issues/131)) ([eb9fc8b](https://www.github.com/googleapis/python-security-private-ca/commit/eb9fc8b1a324505418ece9636e91d844e11845de))

### [1.0.6](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.5...v1.0.6) (2021-09-30)


### Bug Fixes

* improper types in pagers generation ([f86ec89](https://www.github.com/googleapis/python-security-private-ca/commit/f86ec89f3c4537556188064606be005ee7feb056))

### [1.0.5](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.4...v1.0.5) (2021-09-24)


### Bug Fixes

* add 'dict' annotation type to 'request' ([49b5c9a](https://www.github.com/googleapis/python-security-private-ca/commit/49b5c9ae54c594abf1a8158506e2a1ddb6dce67d))

### [1.0.4](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.3...v1.0.4) (2021-08-10)


### Documentation

* **samples:** add local generation for crypto keys ([#98](https://www.github.com/googleapis/python-security-private-ca/issues/98)) ([0668ffd](https://www.github.com/googleapis/python-security-private-ca/commit/0668ffde892bec99a4cd574bbc257fcc2de6c1c7))


### Miscellaneous Chores

* release as 1.0.4 ([#100](https://www.github.com/googleapis/python-security-private-ca/issues/100)) ([47fb407](https://www.github.com/googleapis/python-security-private-ca/commit/47fb4075db02e5c3eaf4f25f3d032a6c2514afce))

### [1.0.3](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.2...v1.0.3) (2021-07-29)


### Bug Fixes

* enable self signed jwt for grpc ([#91](https://www.github.com/googleapis/python-security-private-ca/issues/91)) ([674dd85](https://www.github.com/googleapis/python-security-private-ca/commit/674dd8595b9165fec92097d5bb168357ac7ab1ee))


### Documentation

* add Samples section to CONTRIBUTING.rst ([#84](https://www.github.com/googleapis/python-security-private-ca/issues/84)) ([305dc83](https://www.github.com/googleapis/python-security-private-ca/commit/305dc83eec5215fd84a63e5786d8c93b03c468b8))
* **samples:** private CA python samples ([d753667](https://www.github.com/googleapis/python-security-private-ca/commit/d753667bc3a193931893260cec33d0c68ab621d5))


### Miscellaneous Chores

* release as 1.0.3 ([#92](https://www.github.com/googleapis/python-security-private-ca/issues/92)) ([6026929](https://www.github.com/googleapis/python-security-private-ca/commit/6026929efe36ecec40afbc442f09df609b7c42a8))

### [1.0.2](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.1...v1.0.2) (2021-07-20)


### Bug Fixes

* **deps:** pin 'google-{api,cloud}-core', 'google-auth' to allow 2.x versions ([#83](https://www.github.com/googleapis/python-security-private-ca/issues/83)) ([cd5390c](https://www.github.com/googleapis/python-security-private-ca/commit/cd5390cf5fff50419b000c71431d8ede0de35833))

### [1.0.1](https://www.github.com/googleapis/python-security-private-ca/compare/v1.0.0...v1.0.1) (2021-07-16)


### Bug Fixes

* correct response type of DeleteCaPool ([13e54bf](https://www.github.com/googleapis/python-security-private-ca/commit/13e54bf5ad66f85f1e2165b2cf67604af50ccd0c))
* make allow_config_based_issuance bool optional ([#80](https://www.github.com/googleapis/python-security-private-ca/issues/80)) ([13e54bf](https://www.github.com/googleapis/python-security-private-ca/commit/13e54bf5ad66f85f1e2165b2cf67604af50ccd0c))
* make allow_csr_based_issuance bool optional ([13e54bf](https://www.github.com/googleapis/python-security-private-ca/commit/13e54bf5ad66f85f1e2165b2cf67604af50ccd0c))
* make publish_ca_cert bool optional ([13e54bf](https://www.github.com/googleapis/python-security-private-ca/commit/13e54bf5ad66f85f1e2165b2cf67604af50ccd0c))
* make publish_crl bool optional ([13e54bf](https://www.github.com/googleapis/python-security-private-ca/commit/13e54bf5ad66f85f1e2165b2cf67604af50ccd0c))

## [1.0.0](https://www.github.com/googleapis/python-security-private-ca/compare/v0.4.0...v1.0.0) (2021-07-12)


### Features

* bump release level to production/stable ([#60](https://www.github.com/googleapis/python-security-private-ca/issues/60)) ([170f9be](https://www.github.com/googleapis/python-security-private-ca/commit/170f9be92448278064fd58f2a9302ca2f8c43b04))


### Documentation


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")


nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
]

@nox.session(python=['3.6', '3.7', '3.8', '3.9', '3.10'])
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'asyncmock', 'pytest-asyncio')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/security/privateca_v1/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python='3.9')
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=['3.6', '3.7', '3.8', '3.9'])
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python='3.9')
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx<3.0.0", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )
